src/common/static_ptr.h

```diff
@@ -0,0 +1,409 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include "common/backport14.h"
+
+namespace ceph {
+// `static_ptr`
+// ===========
+//
+// It would be really nice if polymorphism didn't require a bunch of
+// mucking about with the heap. So let's build something where we
+// don't have to do that.
+//
+namespace _mem {
+
+// This, an operator function, is one of the canonical ways to do type
+// erasure in C++ so long as all operations can be done with subsets
+// of the same arguments (which is not true for function type erasure)
+// it's a pretty good one.
+enum class op {
+  copy, move, destroy, size
+};
+template<typename T>
+static std::size_t op_fun(op oper, void* p1, void* p2) noexcept
+{
+  auto me = static_cast<T*>(p1);
+
+  switch (oper) {
+  case op::copy:
+    // One conspicuous downside is that immovable/uncopyable functions
+    // kill compilation right here, even if nobody ever calls the move
+    // or copy methods. Working around this is a pain, since we'd need
+    // four operator functions and a top-level class to
+    // provide/withhold copy/move operations as appropriate.
+    new (p2) T(*me);
```

Comment:
i'm not sure it's reasonable to require move+copy support from all of our types. could we explore some workarounds?

Reply:
I'd like to proceed quickly to a working model.  If the current generality is the problem, I'm in favor of reducing scope.

Reply:
@cbodley I'd rather just fix the two safety issues you pointed out (doing that now) and get this version in. The more general version can be done with a template switch+inheritance, but I'd rather do that in a second patch just so Matt can have at it and we can get some feedback before complicating things.